### PR TITLE
Fix translation of device execution built-ins

### PR DESCRIPTION
**Pull request by AlexeySotkin**
* Change SPIR-V Writer/Reader to consume/produce LLVM IR with blocks and
device side enqueue built-ins as it is produced by clang
* Improve function pointers removal related to OpenCL 2.0 blocks
* Improve tests
* Fix translation of literal structs

### CODE REVIEW

File: lib/SPIRV/SPIRVLowerOCLBlocks.cpp [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+    auto *GVType = dyn_cast<PointerType>(GV.getType()->getElementType());
+    if (GVType && GVType->getElementType()->isFunctionTy())
+      FuncPtrGlbs.push_back(&GV);
+  }
+}
+
+class SPIRVLowerOCLBLocks : public ModulePass {
+
+public:
+  SPIRVLowerOCLBLocks() : ModulePass(ID) {}
```
Comment (neildhickey) [inline lib/SPIRV/SPIRVLowerOCLBlocks.cpp]: The l of Blocks is capitalized

File: lib/SPIRV/CMakeLists.txt [hunk 2]
```diff
@@ -27,6 +27,7 @@ add_llvm_library(LLVMSPIRVLib
   SPIRVLowerConstExpr.cpp
   SPIRVLowerSPIRBlocks.cpp
   SPIRVLowerMemmove.cpp
+  SPIRVLowerOCLBlocks.cpp
```
Comment (neildhickey) [inline lib/SPIRV/CMakeLists.txt]: Could we re-alphabetise this list, I'm sorry I'm a bit of a pedant :)

File: lib/SPIRV/OCL20ToSPIRV.cpp [hunk 3]
```diff
@@ -236,7 +237,6 @@ class OCL20ToSPIRV : public ModulePass, public InstVisitor<OCL20ToSPIRV> {
   void visitCallVecLoadStore(CallInst *CI, StringRef MangledName,
                              const std::string &DemangledName);
 
-  /// Transforms get_mem_fence built-in to SPIR-V function and aligns result
```
Comment (neildhickey) [inline lib/SPIRV/OCL20ToSPIRV.cpp]: Should this line have been deleted?

Comment (AlexeySotkin) [reply lib/SPIRV/OCL20ToSPIRV.cpp]: No. It's a mistake.

File: lib/SPIRV/CMakeLists.txt (hunk 2 above)
Comment (AlexeySotkin) [reply lib/SPIRV/CMakeLists.txt]: No problem :)

**Comment by AlexeySotkin** [comment]
@neildhickey, I have applied your comments. Please take a look. Thanks.

**Comment by neildhickey** [comment]
LGTM

**Pull request merged**